### PR TITLE
Retry when topic leader not available

### LINKED ISSUE #15: Bug: Destination connector stops upon handling first message
### Bug description

Right after starting a pipeline with two `builtin:kafka` connectors (source from one topic, output to another), the destination connector immediately stops:

```
2022-03-25T18:36:56+00:00 DBG starting pipeline component=pipeline.Service pipeline_id=9155d960-165d-450e-8137-3c923b2b018c request_id=5e20b661-d430-4465-b20c-f8e8722a5359
2022-03-25T18:36:56+00:00 INF pipeline started component=pipeline.Service pipeline_id=9155d960-165d-450e-8137-3c923b2b018c request_id=5e20b661-d430-4465-b20c-f8e8722a5359
2022-03-25T18:36:56+00:00 DBG starting destination connector plugin component=connector.Destination connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a
2022-03-25T18:36:56+00:00 DBG starting source connector plugin component=connector.Source connector_id=0f08dab9-3b6c-46e2-b5c1-68513fd9ba24
2022-03-25T18:36:56+00:00 INF request processed duration=0.321186 grpc_method=/api.v1.PipelineService/StartPipeline grpc_status_code=OK http_endpoint="POST /v1/pipelines/9155d960-165d-450e-8137-3c923b2b018c/start" request_id=5e20b661-d430-4465-b20c-f8e8722a5359
2022-03-25T18:36:56+00:00 DBG configuring destination connector plugin component=connector.Destination connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a
2022-03-25T18:36:56+00:00 INF Configuring a Kafka Destination... connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a plugin_name=kafka plugin_type=destination
2022-03-25T18:36:56+00:00 DBG configuring source connector plugin component=connector.Source connector_id=0f08dab9-3b6c-46e2-b5c1-68513fd9ba24
2022-03-25T18:36:56+00:00 INF destination connector plugin successfully started component=connector.Destination connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a
2022-03-25T18:36:56+00:00 INF Configuring a Kafka Source... component=plugin connector_id=0f08dab9-3b6c-46e2-b5c1-68513fd9ba24 plugin_name=kafka plugin_type=source
2022-03-25T18:36:56+00:00 INF source connector plugin successfully started component=connector.Source connector_id=0f08dab9-3b6c-46e2-b5c1-68513fd9ba24
2022-03-25T18:36:56+00:00 DBG stopping destination connector plugin component=connector.Destination connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a
2022-03-25T18:36:56+00:00 DBG tearing down destination connector plugin component=connector.Destination connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a
2022-03-25T18:36:56+00:00 INF Tearing down a Kafka Destination... connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a plugin_name=kafka plugin_type=destination
2022-03-25T18:36:56+00:00 INF connector plugin successfully torn down component=connector.Destination connector_id=1507c91d-ba2d-45a7-8222-8ee34e90db7a
```

### Steps to reproduce

:one: Create pipeline
:two: Add source kafka connector:
```
  {
    "type": "TYPE_SOURCE",
    "plugin": "builtin:kafka",
    "pipelineId": "b44b63c2-aa26-4ecb-aec3-5a4ea0e1d479",
    "config": {
      "name": "kafka-in",
      "settings": {
        "servers": "localhost:9092",
        "topic": "new-json-struct",
        "readFromBeginning": "true"
      }
    }
  }
  ```
:three: Add destination kafka connector:
```
{
  "type": "TYPE_DESTINATION",
  "plugin": "builtin:kafka",
  "pipelineId": "b44b63c2-aa26-4ecb-aec3-5a4ea0e1d479",
  "config": {
    "name": "kafka-out",
    "settings": {
      "servers": "localhost:9092",
      "topic": "output-test-hi"
    }
  }
}
```
:four: Start Pipeline - you'll see the destination connector stop. The destination topic was created on the kafka cluster, but no messages were written to it.

Maybe helpful - if I stop and restart conduit, then the message will successfully write to the destination. So I wonder if something got stuck / into a bad state before restarting?

Also - I tried this with the latest conduit `main` branch, along with updating to the latest commit of this connector - still had this issue.

### Version

v0.2.0-nightly.20220325

### PR DESCRIPTION
**Pull request by hariso**
### Description

Fixes #15,.

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-kafka/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: destination.go [hunk 1]
```diff
@@ -53,10 +56,24 @@ func (d *Destination) Open(ctx context.Context) error {
...
 		record.Key.Bytes(),
 		record.Payload.Bytes(),
 	)
+	// this can happen when the topic doesn't exist and the broker has auto-create enabled
+	// we give it some time to process topic metadata and retry
+	if retry && errors.Is(err, kafka.LeaderNotAvailable) {
+		sdk.Logger(ctx).
+			Info().
+			Err(err).
+			Msg("leader for topic unavailable, will retry")
```
Comment (samirketema) [inline destination.go:73]: Nit: would be nice to add the retry time to the log, particularly since it's 15 seconds long

Comment (samirketema) [reply destination.go:73]: Alternatively, we could retry once every second for _up to_ 15 seconds, to help minimize the amount of time waiting. (Or even use a backoff of sorts, but that may be overkill for this fix 😄 )

**Review by samirketema — APPROVED** [review]
Looks good for fixing the issue, but would be nice to have the gradual/periodic retry as mentioned.

As for the surfacing of the error, seems like that needs to be taken care of in the `Write()` callsite in the main conduit repo. 🙏

**Review by samirketema — APPROVED** [review]
Love the retry logic! 🙌🏾  Thank you so much @hariso ✨

**Comment by hariso** [comment]
> Love the retry logic! 🙌🏾 Thank you so much @hariso sparkles

You're welcome, Samir!

**Review by lovromazgon — APPROVED** [review]
👍

**Pull request merged**